crates/turborepo-scm/src/manual.rs

```diff
@@ -0,0 +1,210 @@
+    let mut includes = Vec::new();
+    let mut excludes = Vec::new();
+    for pattern in inputs {
+        let pattern = pattern.as_ref();
+        if let Some(exclusion) = pattern.strip_prefix('!') {
+            let glob = Path::new(exclusion)
+                .to_slash()
+                .ok_or_else(|| PathError::invalid_utf8_error(exclusion.as_bytes()))?
+                .into_owned();
+            let g = Glob::new(glob.as_str())
+                .map(|g| g.into_owned())
+                .map_err(|e| e.into_owned())?;
+            excludes.push(g);
+        } else {
+            let glob = Path::new(pattern)
+                .to_slash()
+                .ok_or_else(|| PathError::invalid_utf8_error(pattern.as_bytes()))?
+                .into_owned();
+            let g = Glob::new(glob.as_str())
+                .map(|g| g.into_owned())
+                .map_err(|e| e.into_owned())?;
+            includes.push(g);
+        }
+    }
+    let include_pattern = if includes.is_empty() {
+        None
+    } else {
+        Some(any::<Glob<'static>, _>(includes)?)
+    };
+    let exclude_pattern = if excludes.is_empty() {
+        None
+    } else {
+        Some(wax::any::<Glob<'static>, _>(excludes.into_iter())?)
+    };
+    let walker = walker_builder
+        .follow_links(false)
+        .git_ignore(true)
+        .require_git(false)
+        .hidden(false) // this results in yielding hidden files (e.g. .gitignore)
+        .build();
+    for dirent in walker {
+        let dirent = dirent?;
+        let metadata = dirent.metadata()?;
+        // We need to do this here, rather than as a filter, because the root
+        // directory is always yielded and not subject to the supplied filter.
+        if metadata.is_dir() {
+            continue;
+        }
+        let path = AbsoluteSystemPathBuf::new(dirent.path())?;
+        let relative_path = full_package_path.anchor(&path)?;
+        let relative_path = relative_path.to_unix()?;
+        if let Some(include_pattern) = include_pattern.as_ref() {
+            if !include_pattern.is_match(relative_path.as_str()?) {
+                continue;
+            }
+        }
+        if let Some(exclude_pattern) = exclude_pattern.as_ref() {
+            if exclude_pattern.is_match(relative_path.as_str()?) {
+                continue;
+            }
+        }
+        let hash = git_like_hash_file(&path, &metadata)?;
+        hashes.insert(relative_path, hash);
+    }
+    Ok(hashes)
+}
+
+#[cfg(test)]
+mod tests {
+    use turbopath::{RelativeUnixPath, RelativeUnixPathBuf};
+
+    use super::*;
+
+    fn tmp_dir() -> (tempfile::TempDir, AbsoluteSystemPathBuf) {
+        let tmp_dir = tempfile::tempdir().unwrap();
+        let dir = AbsoluteSystemPathBuf::new(tmp_dir.path().to_path_buf())
+            .unwrap()
+            .to_realpath()
+            .unwrap();
+        (tmp_dir, dir)
+    }
+
+    #[test]
+    fn test_get_package_file_hashes_from_processing_gitignore() {
```

Comment:
not a big deal but it might be nice to use something like [insta](https://insta.rs) or [test_case](https://crates.io/crates/test-case) for these kinds of tests

Reply:
I'm kinda on the fence. I like test_case, Alex has been using it I think, but it interacts a little poorly with debugging shortcuts in VSCode once you have more than a couple cases.

Reply:
yeah and it doesn't interact with rustfmt that well. I gave up on it for some more involved test cases in turborepo-cache